Userland/Libraries/LibGfx/ImageFormats/JPEGLoader.cpp

```diff
@@ -1,1985 +1,129 @@
+    source_manager.skip_input_data = [](j_decompress_ptr context, long num_bytes) {
+        if (num_bytes > static_cast<long>(context->src->bytes_in_buffer)) {
+            context->src->bytes_in_buffer = 0;
+            return;
+        }
+        context->src->next_input_byte += num_bytes;
+        context->src->bytes_in_buffer -= num_bytes;
+    };
+    source_manager.resync_to_restart = jpeg_resync_to_restart;
+    source_manager.term_source = [](j_decompress_ptr) {};
 
-    return is_misc || is_table;
-}
+    cinfo.src = &source_manager;
 
-static ErrorOr<void> handle_miscellaneous_or_table(JPEGStream& stream, JPEGLoadingContext& context, Marker const marker)
-{
-    if (is_app_marker(marker)) {
-        TRY(read_app_marker(stream, context, marker - JPEG_APPN0));
-        return {};
+    jpeg_save_markers(&cinfo, JPEG_APP0 + 2, 0xFFFF);
+    if (jpeg_read_header(&cinfo, TRUE) != JPEG_HEADER_OK) {
+        jpeg_destroy_decompress(&cinfo);
+        return Error::from_string_literal("Failed to read JPEG header");
     }
 
-    switch (marker) {
-    case JPEG_COM:
-    case JPEG_DAC:
-        dbgln_if(JPEG_DEBUG, "TODO: implement marker \"{:x}\"", marker);
-        if (auto result = skip_segment(stream); result.is_error()) {
-            dbgln_if(JPEG_DEBUG, "Error skipping marker: {:x}!", marker);
-            return result.release_error();
-        }
-        break;
-    case JPEG_DHT:
-        TRY(read_huffman_table(stream, context));
-        break;
-    case JPEG_DQT:
-        TRY(read_quantization_table(stream, context));
-        break;
-    case JPEG_DRI:
-        TRY(read_restart_interval(stream, context));
-        break;
-    default:
-        dbgln("Unexpected marker: {:x}", marker);
-        VERIFY_NOT_REACHED();
+    if (cinfo.jpeg_color_space == JCS_CMYK || cinfo.jpeg_color_space == JCS_YCCK) {
+        cinfo.out_color_space = JCS_CMYK;
+    } else {
+        cinfo.out_color_space = JCS_EXT_BGRX;
     }
 
-    return {};
-}
-
-static ErrorOr<void> parse_header(JPEGStream& stream, JPEGLoadingContext& context)
-{
-    auto marker = TRY(read_marker_at_cursor(stream));
-    if (marker != JPEG_SOI) {
-        dbgln_if(JPEG_DEBUG, "SOI not found: {:x}!", marker);
-        return Error::from_string_literal("SOI not found");
-    }
-    for (;;) {
-        marker = TRY(read_marker_at_cursor(stream));
+    jpeg_start_decompress(&cinfo);
 
-        if (is_miscellaneous_or_table_marker(marker)) {
-            TRY(handle_miscellaneous_or_table(stream, context, marker));
-            continue;
+    if (cinfo.out_color_space == JCS_EXT_BGRX) {
+        bitmap = TRY(Gfx::Bitmap::create(Gfx::BitmapFormat::BGRx8888, { static_cast<int>(cinfo.output_width), static_cast<int>(cinfo.output_height) }));
+        while (cinfo.output_scanline < cinfo.output_height) {
+            auto* row_ptr = (u8*)bitmap->scanline(cinfo.output_scanline);
+            jpeg_read_scanlines(&cinfo, &row_ptr, 1);
```

Comment:
FYI, if you don't want to regress 12-bit jpeg support, you need to check for bpp and call the 12-bit variant here, similar to https://github.com/libjpeg-turbo/libjpeg-turbo/commit/e8b40f3c2ba187ba95c13c3e8ce21c8534256df7#diff-70a4ed14d4f5eef65a5691a6ec44afe2a836a9d79f1b2d64de2985b614ca3e0cR724

(You might not care about 12bpp jpegs; they're very rare.)

Reply:
Ah that's good to know! Thank you :)

Reply:
Oh wait, no other browsers seem to support 12-bit jpegs either. So let's skip them for now :)